### PR TITLE
[2201.3.x] Add report related operations

### PR DESCRIPTION
**Pull request by LakshanSS**
# Description
Add the following operations related to Reports
1. listReports
2. deleteReport
3. runReportSunc
4. runReportAsync
5. listAsyncRunsOfReport
6. getReportInstanceResult

- Fix https://github.com/wso2-enterprise/choreo/issues/18319
- Fix https://github.com/ballerina-platform/ballerina-extended-library/issues/490

One line release note: 
- Add report related operations

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

**Comment by LakshanSS** [comment]
Had an offline chat with the customer.
Need to add few more operations.

**Review by RolandHewage — COMMENTED** [review]
Let's add some test cases and samples if possible. Rest seems fine.

**Comment by LakshanSS** [comment]
> Let's add some test cases and samples if possible. Rest seems fine.

Created https://github.com/wso2-enterprise/choreo/issues/18807 to track this. Will do via another PR.

File: salesforce/client.bal [hunk 1]
```diff
@@ -245,6 +245,73 @@ public isolated client class Client {
...
+    @display {label: "Delete Report"}
+    isolated remote function deleteReport(@display {label: "Report Id"} string reportId) returns error? {
+        string path = utils:prepareUrl([API_BASE_PATH, ANALYTICS, REPORTS, reportId]);
+        return check self.salesforceClient->delete(path);
+    }
+
+    # Runs an instance of a report synchronously.
+    #
+    # + reportId - Report Id
+    # + return - Reports if successful or else `error`
```
Comment (RolandHewage) [inline salesforce/client.bal:270]: Shall we update the return when successful with the proper record returned.

File: salesforce/client.bal [hunk 2]
```diff
@@ -245,6 +245,73 @@ public isolated client class Client {
...
+    isolated remote function runReportSync(@display {label: "Report Id"} string reportId)
+            returns @display {label: "Report Instances"} ReportInstanceResult|error {
+        string path = utils:prepareUrl([API_BASE_PATH, ANALYTICS, REPORTS, reportId]);
+        return check self.salesforceClient->get(path);
+    }
+
+    # Runs an instance of a report asynchronously.
+    #
+    # + reportId - Report Id
+    # + return - Reports if successful or else `error`
```
Comment (RolandHewage) [inline salesforce/client.bal:281]: Shall we update the return when successful with the proper record returned.

File: salesforce/client.bal [hunk 3]
```diff
@@ -245,6 +245,73 @@ public isolated client class Client {
...
+        string path = utils:prepareUrl([API_BASE_PATH, ANALYTICS, REPORTS, reportId, INSTANCES]);
+        http:Request req = new;
+        req.setJsonPayload({});
+        return check self.salesforceClient->post(path, req);
+    }
+
+    # Lists asynchronous runs of a Report.
+    #
+    # + reportId - Report Id
+    # + return - Reports if successful or else `error`
```
Comment (RolandHewage) [inline salesforce/client.bal:294]: Shall we update the return when successful with the proper record returned.

**Review by RolandHewage — COMMENTED** [review]
Added few comments

File: salesforce/client.bal [hunk 4]
```diff
@@ -245,6 +245,73 @@ public isolated client class Client {
...
+    @display {label: "Run Report Asynchronously"}
+    isolated remote function runReportAsync(@display {label: "Report Id"} string reportId)
+            returns @display {label: "Report Instances"} ReportInstance|error {
+        string path = utils:prepareUrl([API_BASE_PATH, ANALYTICS, REPORTS, reportId, INSTANCES]);
+        http:Request req = new;
+        req.setJsonPayload({});
+        return check self.salesforceClient->post(path, req);
```
Comment (SkNuwanTissera) [inline salesforce/client.bal:288]:
```suggestion
        return check self.salesforceClient->post(path, {});
```

File: salesforce/client.bal [hunk 5]
```diff
@@ -245,6 +245,73 @@ public isolated client class Client {
...
+        return check self.salesforceClient->delete(path);
+    }
+
+    # Runs an instance of a report synchronously.
+    #
+    # + reportId - Report Id
+    # + return - Reports if successful or else `error`
+    @display {label: "Run Report Synchronously"}
+    isolated remote function runReportSync(@display {label: "Report Id"} string reportId)
+            returns @display {label: "Report Instances"} ReportInstanceResult|error {
```
Comment (SanduDS) [inline salesforce/client.bal:273]:
```suggestion
            returns @display {label: "Report Instance"} ReportInstance|error {
```
Can't we do like this as `Runs an instance of a report synchronously` ?

File: salesforce/client.bal (hunk 1 above)
Comment (LakshanSS) [reply salesforce/client.bal]: Done

File: salesforce/client.bal (hunk 2 above)
Comment (LakshanSS) [reply salesforce/client.bal]: Done

File: salesforce/client.bal (hunk 3 above)
Comment (LakshanSS) [reply salesforce/client.bal]: Done

File: salesforce/client.bal (hunk 5 above)
Comment (LakshanSS) [reply salesforce/client.bal]: Fixed the display annotation correctly

**Review by RolandHewage — APPROVED** [review]
Minor change requested. Rest is fine. Therefore approving.

File: salesforce/client.bal [hunk 6]
```diff
@@ -245,6 +245,71 @@ public isolated client class Client {
...
+        return check self.salesforceClient->get(path);
+    }
+
+    # Runs an instance of a report asynchronously.
+    #
+    # + reportId - Report Id
+    # + return - ReportInstance if successful or else `error`
+    @display {label: "Run Report Asynchronously"}
+    isolated remote function runReportAsync(@display {label: "Report Id"} string reportId)
+            returns @display {label: "Report Instances"} ReportInstance|error {
```
Comment (RolandHewage) [inline salesforce/client.bal:284]: @display {label: "Report Instance"}

Instead of

@display {label: "Report Instances"}

Comment (LakshanSS) [reply salesforce/client.bal]: Done

**Review by RolandHewage — APPROVED** [review]

**Review by SanduDS — APPROVED** [review]

File: salesforce/client.bal (hunk 5 above)
Comment (SanduDS) [reply salesforce/client.bal]: ReportInstanceResult --> ReportInstance I guessed, Let's keep as it is for now

Comment (LakshanSS) [reply salesforce/client.bal]: This was named purposefully.

`ReportInstance` is used for another operation `runReportAsync`. That record will only have few fields about the instance.
But `ReportInstanceResult` will have the full detailed result. Thats why named like this.

**Pull request merged**